File: fcfs.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void sjf_scheduling(int n, vector<int> &bt)
{
    vector<int> ct(n), tat(n), wt(n);
    int total_tat = 0, total_wt = 0;

    // Sort the processes based on burst time (Shortest Job First)
    vector<int> sorted_index(n);
    for (int i = 0; i < n; i++)
    {
        sorted_index[i] = i;
    }
    sort(sorted_index.begin(), sorted_index.end(), [&bt](int a, int b)
         { return bt[a] < bt[b]; });

    // Calculate Completion Time (CT)
    ct[sorted_index[0]] = bt[sorted_index[0]]; // First process
    for (int i = 1; i < n; i++)
    {
        ct[sorted_index[i]] = ct[sorted_index[i - 1]] + bt[sorted_index[i]];
    }

    // Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for (int i = 0; i < n; i++)
    {
        tat[i] = ct[i] - 0;     // Assuming Arrival Time is always 0
        wt[i] = tat[i] - bt[i]; // WT = TAT - BT
        total_tat += tat[i];
        total_wt += wt[i];
    }

    // Print the results
    cout << "\nP  | BT | CT | TAT | WT\n";
    for (int i = 0; i < n; i++)
    {
        cout << "P" << i + 1 << " | " << bt[i] << "  | " << ct[i] << "  | " << tat[i] << "   | " << wt[i] << endl;
    }

    // Calculate and print averages
    cout << "\nAvg TAT: " << (float)total_tat / n << endl;
    cout << "Avg WT: " << (float)total_wt / n << endl;
}

int main()
{
    int n;

    cout << "Enter the number of processes: ";
    cin >> n;

    vector<int> bt(n);

    // Input burst time (BT) for each process
    for (int i = 0; i < n; i++)
    {
        cout << "Enter Burst Time for Process P" << i + 1 << ": ";
        cin >> bt[i];
    }

    sjf_scheduling(n, bt);

    return 0;
}

File: fcfs.py

def fcfs_scheduling():
    # Number of processes
    n = int(input("Enter the number of processes: "))
    
    # Lists to store the arrival time and burst time of each process
    processes = []
    
    # Getting input for arrival time and burst time for each process
    for i in range(n):
        at = int(input(f"Enter Arrival Time for Process P{i+1}: "))
        bt = int(input(f"Enter Burst Time for Process P{i+1}: "))
        processes.append({'arrival_time': at, 'burst_time': bt})
    
    # Initialize lists for completion time, turnaround time, and waiting time
    completion_time = [0] * n
    turnaround_time = [0] * n
    waiting_time = [0] * n
    
    # First process completes at its arrival time + burst time (since it starts at time 0)
    completion_time[0] = processes[0]['arrival_time'] + processes[0]['burst_time']
    
    # CT
    for i in range(1, n):
        # Completion time is either the previous process's completion time, or the current process's arrival time
        # whichever is greater, then add the current process's burst time
        completion_time[i] = max(completion_time[i-1], processes[i]['arrival_time']) + processes[i]['burst_time']
    
    # (TAT = CT - AT)
    for i in range(n):
        turnaround_time[i] = completion_time[i] - processes[i]['arrival_time']
    
    # (WT = TAT - BT)
    for i in range(n):
        waiting_time[i] = turnaround_time[i] - processes[i]['burst_time']
    
    # averages
    avg_tat = sum(turnaround_time) / n
    avg_wt = sum(waiting_time) / n
    
    # Print the results
    print("\nProcess | Arrival Time | Burst Time | Completion Time | Turnaround Time | Waiting Time")
    for i in range(n):
        print(f"P{i+1}     | {processes[i]['arrival_time']}           | {processes[i]['burst_time']}         | {completion_time[i]}              | {turnaround_time[i]}              | {waiting_time[i]}")
    
    # Print the average
    print(f"\nAverage Turnaround Time (TAT): {avg_tat:.2f}")
    print(f"Average Waiting Time (WT): {avg_wt:.2f}")

# Call the function to run the program
fcfs_scheduling()
